plugin/core/rpc.py

```diff
@@ -118,39 +185,84 @@ def send_request(
             error_handler: Optional[Callable[[Any], None]] = None,
     ) -> None:
-        self.request_id += 1
         if self.transport is not None:
-            self.logger.outgoing_request(self.request_id, request.method, request.params, blocking=False)
-            self._response_handlers[self.request_id] = (handler, error_handler)
-            self.send_payload(request.to_payload(self.request_id))
+            with self._sync_request_cvar:
+                self.request_id += 1
+                self.logger.outgoing_request(self.request_id, request.method, request.params, blocking=False)
+                self._response_handlers[self.request_id] = (handler, error_handler)
+                self.send_payload(request.to_payload(self.request_id))
         else:
             debug('unable to send', request.method)
             if error_handler is not None:
                 error_handler(None)
             return None
 
-    def execute_request(self, request: Request, timeout: float = DEFAULT_SYNC_REQUEST_TIMEOUT) -> Optional[Any]:
+    def execute_request(
+            self,
+            request: Request,
+            handler: Callable[[Optional[Any]], None],
+            error_handler: Optional[Callable[[Any], None]] = None,
+            timeout: float = DEFAULT_SYNC_REQUEST_TIMEOUT
+    ) -> None:
         """
         Sends a request and waits for response up to timeout (default: 1 second), blocking the current thread.
         """
         if self.transport is None:
             debug('unable to send', request.method)
             return None
 
-        self.request_id += 1
-        request_id = self.request_id
-        self.logger.outgoing_request(request_id, request.method, request.params, blocking=True)
-        self.send_payload(request.to_payload(request_id))
-        result = None
-        try:
-            with self._sync_request_cvar:
+        result = None  # type: Any
+        error = None  # type: Optional[Dict[str, Any]]
+        exception = None  # type: Optional[Exception]
+        with self._sync_request_cvar:
+            try:
+                self.request_id += 1
+                request_id = self.request_id
+                self.logger.outgoing_request(request_id, request.method, request.params, blocking=True)
+                self._sync_request_result.prepare(request_id)  # After this, is_requesting() returns True.
+                self.send_payload(request.to_payload(request_id))
                 # We go to sleep. We wake up once another thread calls .notify() on this condition variable.
-                self._sync_request_cvar.wait_for(lambda: request_id in self._sync_request_results, timeout)
-                result = self._sync_request_results.pop(request_id)
-        except KeyError:
-            debug('timeout on', request.method)
-            return None
-        return result
+                self._sync_request_cvar.wait_for(self._sync_request_result.is_ready, timeout)
```

Comment:
With some frantic testing I was able to trigger timeout and then things went in a bit unexpected way because timeout doesn't trigger `KeyError` exception. Instead, it returns `False` here. So I would suggest something like this:
```suggestion
                timed_out = self._sync_request_cvar.wait_for(self._sync_request_result.is_ready, timeout) is False
                if timed_out:
                    self._sync_request_result.reset()
                    error = {"message": '{}({}): TIMEOUT'.format(request.method, request_id)}
```